src/member.ts

```diff
@@ -0,0 +1,358 @@
+                const who = req.req.who;
+                this.knownMembers = this.knownMembers.filter(
+                  (id) => id !== who
+                );
+              }
+            }
+            req.finish();
+            if (this.isMember()) {
+              this.emit("block", req);
+            }
+          }
+          this.incrementMemberIndexFor(id);
+        } else {
+          continue;
+        }
+      } else {
+        console.warn("Got invalid item", item);
+        this.incrementMemberIndexFor(id);
+      }
+    }
+
+    if (hasProcessed) {
+      const otherRun = this.processFeeds()
+      return hasProcessed || otherRun;
+    } else return hasProcessed;
+  }
+
+  private updateFeedFor(id: ID, feed: FeedItem[]) {
+    this.knownFeeds[id] = feed.slice(0);
+  }
+
+  private hasNewFeedItemsFor(id: ID): boolean {
+    return this.getMemberIndexFor(id) === this.getFeedFor(id).length - 1;
+  }
+
+  private getMemberIndexFor(id: ID): number {
+    if (this.memberIndexes[id]) return this.memberIndexes[id];
+    this.memberIndexes[id] = 0;
+    return 0;
+  }
+
+  private incrementMemberIndexFor(id: ID) {
+    this.memberIndexes[id] = this.getMemberIndexFor(id) + 1;
+  }
+
+  private getFeedFor(id: ID): FeedItem[] {
+    if (!this.knownFeeds[id]) {
+      this.knownFeeds[id] = [];
+    }
+
+    return this.knownFeeds[id];
+  }
+
+  private hasRequest(id: ID): boolean {
+    return !!this.requests[id];
+  }
+
+  private trackRequest(request: Request) {
+    const { id } = request;
+    this.requests[id] = new RequestState(request);
+  }
+
+  private getRequest(id: ID): RequestState {
+    return this.requests[id];
+  }
+
+  requestAdd(who: ID): Request {
+    return this.makeRequest(who, "add");
+  }
+
+  requestRemove(who: ID): Request {
+    return this.makeRequest(who, "remove");
+  }
+
+  makeRequest(who: ID, operation: Operation): Request {
+    const req = {
+      type: REQUEST_TYPE,
+      id: makeID(),
+      from: this.id,
+      who,
+      operation,
+    };
+
+    this.ownFeed.push(req);
+
+    return req;
+  }
+
+  makeResponse(request: Request, response: ResponseType): Response {
+    const { id, from } = request;
+    const res = {
+      type: RESPONSE_TYPE,
+      from,
+      id,
+      response,
+    };
+
+    this.ownFeed.push(res);
+
+    return res;
+  }
+
+  acceptRequest(request: Request): Response {
+    return this.makeResponse(request, "accept");
+  }
+
+  denyRequest(request: Request): Response {
+    return this.makeResponse(request, "deny");
+  }
+
+  isMember() {
+    return this.knownMembers.includes(this.id);
+  }
+}
+
+export function isRequest(item: FeedItem): item is Request {
+  return item.type === REQUEST_TYPE;
+}
+
+export function isResponse(item: FeedItem): item is Response {
+  return item.type === RESPONSE_TYPE;
+}
+
+function isMerge(item: Request): item is MergeRequest {
+  return (item as MergeRequest).toMerge !== undefined;
+}
+
+function isWho(item: Request): item is WhoRequest {
```

Comment:
When I first read `isWho` I was pretty confused, same for `WhoRequest` Maybe rename?

Reply:
Renamed